### PR TITLE
Fix .NET Core on Unix handling of TimeZoneInfo as far as possible

### PR DESCRIPTION
**Pull request by jskeet**
This is mostly just a starting point to see if the test failures go down...

(Fixing #1140)

### CODE REVIEW

**Comment by jskeet** [comment]
Travis results of first change:
Failures: 650, errors: 18

So that's at least somewhat better...

**Comment by jskeet** [comment]
When this PR is merged (so I've got a stable link to the final commit) I'll file an issue on the corefx repo.

**Comment by jskeet** [comment]
All tests pass (or are skipped) on Gabriel, Travis and Folio (WSL).

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs [hunk 1]
```diff
@@ -37,15 +40,34 @@ public class BclDateTimeZoneTest
...
-        [TestCaseSource(nameof(BclZones))]
+        [TestCaseSource(nameof(BclZonesAndIds))]
         [Category("BrokenOnMonoLinux")]
-        public void AllZoneTransitions(TimeZoneInfo windowsZone)
+        public void AllZoneTransitions(TimeZoneInfo windowsZone, string id)
         {
             var nodaZone = BclDateTimeZone.FromTimeZoneInfo(windowsZone);
 
+            // Currently .NET Core doesn't expose the information we need to determine any DST recurrence
+            // after the final tzif rule. (Bug to be filed against corefx.) For the moment, limit how far we check.
```
Comment (malcolmr) [inline src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: (is this referring to one of the bugs you already filed?)

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs [hunk 2]
```diff
@@ -67,14 +89,14 @@ public void DisplayName(TimeZoneInfo windowsZone)
         /// This test catches situations where the Noda Time representation doesn't have all the
         /// transitions it should; AllZoneTransitions may pass not spot times when we *should* have
         /// a transition, because it only uses the transitions it knows about. Instead, here we
-        /// check each day between 1st January 1950 and 1st January 2050. We use midnight UTC, but
+        /// check each week between 1st January 1950 and 1st January 2050. We use midnight UTC, but
```
Comment (malcolmr) [inline src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: It's actually 2050 or 2037, isn't it? (Not sure how precise you want to be here.)

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs [hunk 3]
```diff
@@ -110,8 +132,12 @@ private void ValidateZoneEveryWeek(TimeZoneInfo windowsZone)
         {
             var nodaZone = BclDateTimeZone.FromTimeZoneInfo(windowsZone);
 
+            // Currently .NET Core doesn't expose the information we need to determine any DST recurrence
+            // after the final tzif rule. (Bug to be filed against corefx.) For the moment, limit how far we check.
```
Comment (malcolmr) [inline src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: Ditto. Also is this worth extracting to a test helper anywhere?

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs [hunk 4]
```diff
@@ -136,6 +162,13 @@ public void ForSystemDefault()
         [Test]
         public void DateTimeMinValueStartRuleExtendsToBeginningOfTime()
         {
+            // .NET Core on Unix loses data from rules provided to CreateCustomTimeZone :(
+            // (It assumes the rules have been created from tzif files, which isn't the case here.)
+            if (TestHelper.IsRunningOnDotNetCoreUnix)
```
Comment (malcolmr) [inline src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: Is there an assume-like construct in NUnit? That might be clearer than `if (expr) Assert.Ignore()`.

(Also same pattern elsewhere.)

File: src/NodaTime/TimeZones/BclDateTimeZone.cs [hunk 5]
```diff
@@ -92,6 +98,25 @@ public override ZoneInterval GetZoneInterval(Instant instant)
             return new BclDateTimeZone(bclZone, Offset.Min(standardOffset, minRuleOffset), Offset.Max(standardOffset, maxRuleOffset), cachedMap);
         }
 
+        /// <summary>
+        /// .NET Core on Unix adjustment rules can't currently be treated like regular Windows ones.
+        /// Instead of dividing time into periods by year, the rules are read from TZIF files, so are like
+        /// our PrecalculatedDateTimeZone. This is only visibele for testing purposes.
```
Comment (malcolmr) [inline src/NodaTime/TimeZones/BclDateTimeZone.cs]: Typo: visible.

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs (hunk 1 above)
Comment (jskeet) [reply src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: Will update to link to the existing one.

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs (hunk 2 above)
Comment (jskeet) [reply src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: Updated.

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs (hunk 3 above)
Comment (jskeet) [reply src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: I would do if we needed it more than twice, or permanently. As I'm hoping to get rid of this, it didn't feel worth it.

File: src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs (hunk 4 above)
Comment (jskeet) [reply src/NodaTime.Test/TimeZones/BclDateTimeZoneTest.cs]: Not out of the box. It looks like we could do it with some extra code though:
https://amido.com/blog/conditional-ignore-nunit-and-the-ability-to-conditionally-ignore-a-test/

I may have a look if I get some time, but I won't try to do it in this PR.

File: src/NodaTime/TimeZones/BclDateTimeZone.cs (hunk 5 above)
Comment (jskeet) [reply src/NodaTime/TimeZones/BclDateTimeZone.cs]: Thanks :)

**Pull request merged**